components/raftstore/src/store/fsm/apply.rs

```diff
@@ -3175,13 +3158,24 @@ where
         fail_point!("on_handle_apply_2", self.delegate.id() == 2, |_| {});
         fail_point!("on_handle_apply", |_| {});
 
-        if apply.entries.is_empty() || self.delegate.pending_remove || self.delegate.stopped {
+        if apply.entries.range.is_empty() || self.delegate.pending_remove || self.delegate.stopped {
             return;
         }
 
+        let mut entries = apply.entries.take_entries();
+        if entries.is_empty() {
+            let rid = self.delegate.region_id();
+            let StdRange { start, end } = apply.entries.range;
+            entries = Vec::with_capacity((end - start) as usize);
+            self.delegate
+                .raft_engine
+                .fetch_entries_to(rid, start, end, None, &mut entries)
```

Comment:
why the entries are empty but the range is not.

Reply:
When cached raft entries are deleted because of no memory, fetched committed entries will also be cleaned to release memory.

Reply:
Will it affect the performance?

Reply:
Yes because it drops entry cache but load entries from storage. We can't avoid it.